### PR TITLE
Fix analytics component

### LINKED ISSUE #460: Pageviews not sent when navigating to a link with the same title
**_Bug report_**

Pageview are not sent when you navigate from one page to another with the same title if the second one has the property `isReady` equals `true` (i.e. it was already fetched).

This happen, for example, if you use the `@frontity/head-tags` package, as it set the same title for different archive pages.

- I've updated both my project and my packages to the latest Frontity versions: https://docs.frontity.org/guides/keep-frontity-updated
- I've searched for existing issues.
- I've checked the documentation: https://docs.frontity.org
- I've checked the community forum: https://community.frontity.org

**Expected behavior**

To send a pageview for the page you are navigating to.

**Observed behavior**

No pageview is sent.

**Steps involved to reproduce the problem**

- Go to frontity.org/blog
- Click on the next page link (frontity.org/blog/page/2)
- Go back to the previous page
- A pageview shoud have been sent when going back, but it hasn't.

**Possible solution**

I was looking at the code and it looks like we can remove the title check that we are doing in the `roots.analytics` component as it's kind of redundant code.

Simply remove the next lines

https://github.com/frontity/frontity/blob/bb26a7ce8075b623ba6daf9af68306844193f9da/packages/analytics/src/components/index.tsx#L14-L15

and the comparison between `prevTitle` and `title` here, along with the `setPrevTitle(title)` code.

https://github.com/frontity/frontity/blob/bb26a7ce8075b623ba6daf9af68306844193f9da/packages/analytics/src/components/index.tsx#L34-L36

### PR DESCRIPTION
**Pull request by DAreRodz**
**What**:

It fixes the `@frontity/analytics` root component that wasn't sending pageviews when navigating to a new link with the same title.

I also take the opportunity to refactor the type definitions for the `@frontity/analytics` package and match what was discussed here in the community: [TypeScript: use two different types](https://community.frontity.org/t/typescript-use-two-different-types/1681)

Fixes #460 

**Why**:

To fix a bug that's causing some pageviews not to be sent.

**How**:

Just removing some innecessary checks on a page title that we were doing in the `@frontity/analytics` root component.

**Tasks**:

- [x] Code
- [x] TypeScript
- [x] End to end tests
- [x] Changeset (with link to its [Feature Discussion](https://community.frontity.org/c/33) if it exists)

**Unrelated Tasks**

- TSDocs
- Unit tests
- TypeScript tests
- Update starter themes
- Update other packages
- Link to PR in [Documentation](https://github.com/frontity/gitbook-docs/)
- Community discussions

### CODE REVIEW

**Review by luisherranz — CHANGES_REQUESTED** [review]
While trying to simplify the types, we discovered a bug with the types of `@frontity/google-analytics`.

This was the conversation: https://www.youtube.com/watch?v=WKTXO9ziZ6g

**Comment by DAreRodz** [comment]
@luisherranz, I've changed my mind. Could you push your changes?

I think I've made it work but I don't know what I did differently.  :sweat_smile:

**Comment by luisherranz** [comment]
Done 👍

**Comment by DAreRodz** [comment]
Thanks, I'll take a look.

**Comment by DAreRodz** [comment]
I didn't want to write such a long comment. I hope everything is understandable and clear. Sorry if it's not, @luisherranz, @michalczaplinski. 😅

**TL;DR**

- Do not use generics for exported code that is mean to be imported by other packages
- Avoid exporting types that are only used internaly (the `[key: string]` part)
- When an action could be reimplemented, try to pass the less possible amount of packages to that action type so it's more difficult to have a type conflict.

---

Ok, I guess the problem we had with TypeScript here was an edge case related to the way we implemented the analytics packages. Also, I don't clearly understand why TypeScript doesn't complain when you define the types this way but it does when you try to implement code using those types.

Well, let me try to explain it. 😄

The current implementation of the analytics package types has defined two optional `sendPageview` and `sendEvent` actions for any namespace (because we don't know the name of all analytics packages).

https://github.com/frontity/frontity/blob/5baa5094124f27aa711fc667227b66bfd0be5599/packages/analytics/types.ts#L24-L27

These types are exported in the `Analytics` interface, which is a generic type that accepts one argument for all the packages that is passed down to actions as the `Pkgs` argument you see above. By default, `Pkgs` is:

https://github.com/frontity/frontity/blob/5baa5094124f27aa711fc667227b66bfd0be5599/packages/analytics/types.ts#L36

The problem start when we define the `GoogleAnalytics` types.

Looking at the code at it is right now, we are using `Analytics` without passing any argument when defining `actions` types

https://github.com/frontity/frontity/blob/5baa5094124f27aa711fc667227b66bfd0be5599/packages/google-analytics/types.ts#L34-L39

The `Packages` type we are passing to actions is:

https://github.com/frontity/frontity/blob/5baa5094124f27aa711fc667227b66bfd0be5599/packages/google-analytics/types.ts#L42

At this point, I'm going to write some snippets to clarify what happen. These are the final types

```ts
type AnalyticsPkgs = Analytics & Router & Source; 
type GoogleAnalyticsPkgs = GoogleAnalytics & Analytics<GoogleAnalytics>;

interface GoogleAnalytis {
  actions: {
    analytics: {
      sendPageview: Action<AnalyticsPkgs, Pageview>; 
      sendEvent: Action<AnalyticsPkgs, Event>; 
    };
    googleAnalytics: { 
      sendPageview: Action<GoogleAnalyticsPkgs, Pageview>; 
      sendEvent: Action<GoogleAnalyticsPkgs, GoogleAnalyticsEvent>; 
    };
    [key: string]: {
      sendPageview: Action<AnalyticsPkgs, Pageview>; 
      sendEvent: Action<AnalyticsPkgs, Event>; 
    };
  }; 
}
```

As it can be seen in the code above, `sendPageview` and `sendEvent` actions inside `googleAnalytics` namespace would receive a state derived from `GoogleAnalyticsPkgs`, which conflicts with the state received by actions in `[key: string]`, which is derived from `AnalyticsPkgs`. They conflict because one has types from `GoogleAnalytics` and the other has types from `Router` and `Source`.

The error that TypeScript gives (and I don't understand why it's given when you implement the code and not when you define the types) is

```
Property 'googleAnalytics' is incompatible with index signature.
    Types of property 'sendPageview' are incompatible.
        Types of parameters '__0' and '__0' are incompatible.
            Types of property 'state' are incompatible.
                Property 'googleAnalytics' is missing in type...
```

If we would instead have define the actions passing `GoogleAnalytics` packages as argument to the generic `Analytics` type like this:

```ts
interface GoogleAnalytics extends Package {
  actions: Analytics<Packages>["actions"] & {
    googleAnalytics: {
      sendPageview: Action<Packages, Pageview>;
      sendEvent: Action<Packages, GoogleAnalyticsEvent>;
    };
  };
}
```

The resultant types would have been

```ts
type AnalyticsPkgs = Analytics & Router & Source; 
type GoogleAnalyticsPkgs = GoogleAnalytics & Analytics<GoogleAnalytics>;

interface GoogleAnalytis {
  actions: {
    analytics: {
      sendPageview: Action<GoogleAnalyticsPkgs, Pageview>; 
      sendEvent: Action<GoogleAnalyticsPkgs, Event>; 
    };
    googleAnalytics: { 
      sendPageview: Action<GoogleAnalyticsPkgs, Pageview>; 
      sendEvent: Action<GoogleAnalyticsPkgs, GoogleAnalyticsEvent>; 
    };
    [key: string]: {
      sendPageview: Action<GoogleAnalyticsPkgs, Pageview>; 
      sendEvent: Action<GoogleAnalyticsPkgs, Event>; 
    };
  }; 
}
```

Types are fine now, but after trying this it's still failing! Why?

Because the `analytics` code that we are importing here

https://github.com/frontity/frontity/blob/1fc5c4efe1ed737d75ae9c126300514e92674422/packages/google-analytics/src/index.ts#L1

and its actions that we are adding here

https://github.com/frontity/frontity/blob/1fc5c4efe1ed737d75ae9c126300514e92674422/packages/google-analytics/src/index.ts#L14

have the default types for `Packages` (i.e. `Analytics & Router & Source`), and so it receives a different `state`.

This is the error now:

```
The types of 'analytics.sendPageview' are incompatible between these types.
    Types of parameters '__0' and '__0' are incompatible.
        Types of property 'state' are incompatible.
            Type 'ResolveState<...>' is missing the following properties from type 'ResolveState<...>': router, source
```

---

These are the conclusions I've extracted:

- Do not use generics for exported code that is mean to be imported by other packages.
- Avoid exporting types that are only used internaly (the `[key: string]` part).
- When an action could be reimplemented, try to pass the less possible amount of packages to that action type so it's more difficult to have a type conflict.

**Comment by michalczaplinski** [comment]
@DAreRodz  Just testing integromat because the scenario stopped working. Let me know if you get it please 🙁

**Review by luisherranz — APPROVED** [review]
Hey, sorry it took me so long to do this review. Being a critical bug I should've given it much more priority.

@SantosGuillamot feel free to merge and do a release.
